tools/include/eosio/abigen.hpp

```diff
@@ -799,10 +800,72 @@ namespace eosio { namespace cdt {
             }
             return true;
          }
+
+         const clang::CXXRecordDecl* find_contract_class(const clang::ASTContext &ctx) const {
+            const auto* translation_unit = ctx.getTranslationUnitDecl();
+            // scanning entire translation unit to find contract class
+            for (const clang::Decl* cur_decl : translation_unit->decls()) {
+               if (const auto* cxx_decl = llvm::dyn_cast<clang::CXXRecordDecl>(cur_decl)) {
+                  
+                  if (cxx_decl->isEosioContract()) {
+                     auto attr_name = cxx_decl->getEosioContractAttr()->getName().str();
+                     auto name = attr_name.empty() ? cxx_decl->getName().str() : attr_name;
+                     if (name == ag.get_contract_name())
+                        return cxx_decl;
+                  }
+                  else {
+                     const auto* parent_decl = llvm::dyn_cast<clang::CXXRecordDecl>(cxx_decl->getParent());
+                     if (parent_decl && parent_decl->isEosioContract()) {
+                        auto attr_name = parent_decl->getEosioContractAttr()->getName().str();
+                        auto name = attr_name.empty() ? parent_decl->getName().str() : attr_name;
+                        if (name == ag.get_contract_name())
+                           return parent_decl;
+                     }
+                  }
+               }
+            }
+
+            return nullptr;
+         }
+
+         bool is_same_type(const clang::Decl* decl1, const clang::CXXRecordDecl* decl2) const {
+            if (!decl1 || !decl2)
+               return false;
+            if (decl1 == decl2)
+               return true;
+            
+            // checking if declaration is a typedef or using
+            if (const clang::TypedefNameDecl* typedef_decl = llvm::dyn_cast<clang::TypedefNameDecl>(decl1)) {
+               if (const auto* cur_type = typedef_decl->getUnderlyingType().getTypePtrOrNull()) {
+                  if (decl2 == cur_type->getAsCXXRecordDecl()) {
+                        return true;
+                  }
+               }
+            }
+
+            return false;
+         }
+         
+         bool defined_in_contract(const clang::ClassTemplateSpecializationDecl* decl) {
+
+            if (!contract_class) {
+               contract_class = find_contract_class(decl->getASTContext());
+               if (!contract_class) 
+                  CDT_INTERNAL_ERROR("contract class not found");
```

Comment:
this is a fatal error. we must have a contract class that corresponds to c++ filename or `--contract` option

Reply:
But will that always be accessible from the current translation unit?

Reply:
good point.. maybe you are right, we should just return `false` if no contract.
So this is abi generator. Currently we are not traversing translation units other than a main file:
https://github.com/AntelopeIO/cdt/blob/extend-abi-non-contract/tools/include/eosio/abigen.hpp#L878

However if we will change that in any way we may want to just return `false`. That is fine, if type is defined in main file, it will be picked up anyway.